File: mobile-app/packages/services/test/pro_upgrade_service_test.dart
import 'dart:convert';
import 'package:smwa_services/services.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:test/test.dart';

class _Flag {
  bool value = false;
}

void main() {
  test('checkoutMock posts and updates store', () async {
    var method = '';
    final client = MockClient((req) async {
      method = req.method;
      expect(req.url.toString(),
          'http://localhost:12111/v1/checkout/sessions');
      return http.Response(jsonEncode({'status': 'open'}), 200);
    });
    final flag = _Flag();
    final svc = ProUpgradeService((v) async => flag.value = v, client: client);
    final ok = await svc.checkoutMock();
    expect(ok, isTrue);
    expect(method, 'POST');
    expect(flag.value, isTrue);
  });

  test('checkoutMock returns false on error', () async {
    final client = MockClient((req) async => http.Response('', 500));
    final flag = _Flag();
    final svc = ProUpgradeService((v) async => flag.value = v, client: client);
    final ok = await svc.checkoutMock();
    expect(ok, isFalse);
    expect(flag.value, isFalse);
  });
}

File: web-app/tests/ProUpgradeService.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { ProUpgradeService } from '../src/services/ProUpgradeService';

describe('ProUpgradeService', () => {
  beforeEach(() => {
    vi.restoreAllMocks();
  });
  afterEach(() => {
    vi.restoreAllMocks();
  });

  it('posts checkout parameters and returns true on success', async () => {
    const fetchMock = vi.fn().mockResolvedValue({ ok: true });
    global.fetch = fetchMock as any;
    const svc = new ProUpgradeService();
    const ok = await svc.checkoutMock();
    expect(ok).toBe(true);
    expect(fetchMock).toHaveBeenCalledWith(
      'http://localhost:12111/v1/checkout/sessions',
      expect.objectContaining({
        method: 'POST',
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      })
    );
    const body = (fetchMock.mock.calls[0][1] as any).body as string;
    expect(body).toContain('mode=payment');
    expect(body).toContain('success_url');
    expect(body).toContain('cancel_url');
  });

  it('returns false on network failure', async () => {
    global.fetch = vi.fn().mockRejectedValue(new Error('net')) as any;
    const svc = new ProUpgradeService();
    const ok = await svc.checkoutMock();
    expect(ok).toBe(false);
  });

  it('returns false for non-ok response', async () => {
    global.fetch = vi.fn().mockResolvedValue({ ok: false }) as any;
    const svc = new ProUpgradeService();
    const ok = await svc.checkoutMock();
    expect(ok).toBe(false);
  });
});
